website/pages/docs/caching-strategies.mdx

```diff
@@ -0,0 +1,281 @@
+---
+title: Caching Strategies
+---
+
+# Caching Strategies
+
+Caching is a core strategy for improving the performance and scalability of GraphQL
+servers. Because GraphQL allows clients to specify exactly what they need, the server often
+does more work per request compared to REST APIs.
+
+This guide explores different levels of caching in a GraphQL.js so you can apply the right
+strategy for your application.
+
+## Why caching matters
+
+GraphQL servers commonly face performance bottlenecks due to repeated fetching of the
+same data, costly resolver logic, or expensive database queries. Since GraphQL shifts
+much of the composition responsibility to the server, caching becomes essential for maintaining
+fast response times and managing backend load.
+
+## Caching levels
+
+There are several opportunities to apply caching within a GraphQL server:
+
+- **Resolver-level caching**: Cache the result of specific fields.
+- **Request-level caching**: Batch and cache repeated access to backend resources
+within a single operation.
+- **Operation result caching**: Reuse the entire response for repeated identical queries.
+- **Schema caching**: Cache the compiled schema when startup cost is high.
+- **Transport/middleware caching**: Leverage caching behavior in HTTP servers or proxies.
+
+Understanding where caching fits in your application flow helps you apply it strategically
+without overcomplicating your system.
+
+## Resolver-level caching
+
+Resolver-level caching is useful when a specific field’s value is expensive to compute and 
+commonly requested with the same arguments. Instead of recomputing or refetching the data on 
+every request, you can store the result temporarily in memory and return it directly when the 
+same input appears again.
+
+### Use cases
+
+- Fields backed by slow or rate-limited APIs
+- Fields that require complex computation
+- Data that doesn't change frequently
+
+For example, consider a field that returns information about a product:
+
+```js
+// utils/cache.js
+import LRU from 'lru-cache';
+
+export const productCache = new LRU({ max: 1000, ttl: 1000 * 60 }); // 1 min TTL
+```
+
+The next example shows how to use that cache inside a resolver to avoid repeated database
+lookups:
+
+```js
+// resolvers/product.js
+import { productCache } from '../utils/cache.js';
+
+export const resolvers = {
+  Query: {
+    async product(_, { id }, context) {
+      const cached = productCache.get(id);
+      if (cached) return cached;
+
+      const product = await context.db.products.findById(id);
+      productCache.set(id, product);
+      return product;
+    },
```

Comment:
This code is vulnerable to race conditions where many requests for the same data can be kicked off in parallel during the initial fetch. By caching the promise rather than the result, concurrent requests to the same data can be cached without introducing a race condition:

```suggestion
    product(_, { id }, context) {
      const cached = productCache.get(id);
      if (cached) return cached;

      const productPromise = context.db.products.findById(id);
      productCache.set(id, productPromise);
      return productPromise;
    },
```